File: app/src/main/java/ru/e6atb/chat/SwipeDismissDecider.java
package ru.e6atb.chat;

final class SwipeDismissDecider {
	private SwipeDismissDecider() {
	}

	static float dragTranslation(float startTranslationPx, float downRawYPx, float currentRawYPx) {
		return Math.max(0.0f, startTranslationPx + currentRawYPx - downRawYPx);
	}

	static boolean shouldDismiss(float distancePx, int heightPx, float velocityYPx, float density) {
		float safeDensity = Math.max(1.0f, density);
		float distanceThreshold = Math.max(96.0f * safeDensity, heightPx * 0.28f);
		boolean farEnough = distancePx >= distanceThreshold;
		boolean fastEnough = distancePx >= 24.0f * safeDensity
				&& velocityYPx >= 900.0f * safeDensity;
		return distancePx > 0.0f && (farEnough || fastEnough);
	}
}

File: app/src/main/kotlin/ru/e6atb/chat/SwipeDismissDecider.kt
package ru.e6atb.chat

object SwipeDismissDecider {
    @JvmStatic fun dragTranslation(startTranslationPx: Float, downRawYPx: Float, currentRawYPx: Float) = maxOf(0f, startTranslationPx + currentRawYPx - downRawYPx)
    @JvmStatic fun animationOffset(startPx: Float, targetPx: Float, progress: Float) = startPx + (targetPx - startPx) * progress.coerceIn(0f, 1f)
    @JvmStatic fun shouldDismiss(distancePx: Float, heightPx: Int, velocityYPx: Float, density: Float): Boolean {
        val safeDensity = maxOf(1f, density)
        return distancePx > 0f && (distancePx >= maxOf(96f * safeDensity, heightPx * .28f) || (distancePx >= 24f * safeDensity && velocityYPx >= 900f * safeDensity))
    }
}
